### PR TITLE
Fix issue #1631, Singleton and single navigation property possible without key defined on its type

### PR DESCRIPTION
**Pull request by xuzhg**
…thout key defined on its type

### Issues

*This pull request fixes issue #1631.*

### Description

* Enable Singleton defined without key 
* Enable Single-Value navigation property defined without key.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.AspNet.OData.Shared/Builder/ODataModelBuilder.cs [hunk 1]
```diff
@@ -608,20 +608,30 @@ public virtual void ValidateModel(IEdmModel model)
...
-                    throw Error.InvalidOperation(SRResources.EntityTypeDoesntHaveKeyDefined, entity.Name);
+                    throw Error.InvalidOperation(SRResources.EntitySetTypeHasNoKeys, entitySet.Name,
+                        entitySet.EntityType().FullName());
                 }
             }
 
-            foreach (IEdmNavigationSource navigationSource in model.EntityContainer.Elements.OfType<IEdmNavigationSource>())
+            // The type of collection navigation property should have key(s) defined.
+            HashSet<IEdmStructuredType> visited = new HashSet<IEdmStructuredType>();
+            foreach (IEdmStructuredType structuredType in model.SchemaElementsAcrossModels().OfType<IEdmStructuredType>())
```
Comment (mikepizzo) [inline src/Microsoft.AspNet.OData.Shared/Builder/ODataModelBuilder.cs]: Is this used anywhere?

Comment (mikepizzo) [reply src/Microsoft.AspNet.OData.Shared/Builder/ODataModelBuilder.cs]: Comment showed up on wrong line; is the HashSet "visited" defined on line 622 used anywhere?

---
In reply to: [227550096](https://github.com/OData/WebApi/pull/1641#discussion_r227550096) [](ancestors = 227550096)

**Comment by mikepizzo** [comment]
>         public virtual void ValidateModel(IEdmModel model)

Is there a reason we do separate model validation in ModelBuilder, rather than just using the EdmValidator?

---
Refers to: src/Microsoft.AspNet.OData.Shared/Builder/ODataModelBuilder.cs:604 in e5a844b. [](commit_id = e5a844be04cfdb747304d7665523bc7c7184687b, deletion_comment = False)

**Comment by xuzhg** [comment]
@mikepizzo I think the main reason is to allow error tolerance. The validation rules by default maybe too heavy for some scenarios. However, when the customer get Edm model by calling `GetEdmModel()`, he can call `model.Validate(out errors)` to verify the model by himself. 
In Web API, we add the minimal validation rules for the model by introducing this separate model validation.

File: src/Microsoft.AspNet.OData.Shared/Builder/ODataModelBuilder.cs (hunk 1 above)
Comment (xuzhg) [reply src/Microsoft.AspNet.OData.Shared/Builder/ODataModelBuilder.cs]: #resolved.

**Pull request merged**